z3/src/ast.rs

```diff
@@ -300,6 +301,20 @@ pub trait Ast<'ctx>: Sized + fmt::Debug + Into<Dynamic<'ctx>> {
     fn is_const(&self) -> bool {
         self.is_app() && self.num_children() == 0
     }
+
+    /// Return the `FuncDecl` of the `Ast`.
```

Comment:
Could you add a note for the `self.kind() == AstKind::App` precondition?

Reply:
Do you mean in the `is_const()` fn (via `self.is_app()`)?  I want to admit I'm mostly just porting from the python/C++ APIs:
```cpp
         bool is_app() const { return kind() == Z3_APP_AST || kind() == Z3_NUMERAL_AST; }
```
and the various Python asserts that an Ast is actually an app:
```python
    def decl(self):
        if z3_debug():
            _z3_assert(is_app(self), "Z3 application expected")
        return FuncDeclRef(Z3_get_app_decl(self.ctx_ref(), self.as_ast()), self.ctx)
```

I suspect that the python assert -- which is only in debug mode! -- is based on python's loose typing, and it may be unnecessary for the Rust library, but I erred on the side of caution.  (It's worth noting that converting from an Z3_ast to Z3_app is just a cast, and the pointer is unchanged.)

In C++, the class hierarchy is:
* `ast`
* `func_decl` subclass of `ast`
* `expr` subclass of `ast`

where `expr` seems most analogous to the Rust `Ast`.

In short, I don't really understand the class hierarchy model, and I'm happy to do whatever others think is safest.

Reply:
I meant just to mention that the method will panic if the ast kind is not AstKind::App

Reply:
Added that message in the doc comment.